### ISSUE TITLE
1 deg CICE grid very slightly misaligned with MOM

### ISSUE THREAD

**Issue by aekiss** [COSIMA/access-om2#279]
The 1deg CICE grid values are very slightly different from MOM in both latitude and longitude for both T and U points - see [this notebook](https://github.com/aekiss/notebooks/blob/e95d7e9c0e91ffd77cdcafd578e8bb14af65e109/grid_check.ipynb).

Perhaps there was loss of precision due to conversion to single precision and back to double?

Possibly related: https://github.com/COSIMA/access-om2/issues/271

**Comment by aekiss** [comment]
would be good to have grid sanity checks to catch this sort of thing https://github.com/COSIMA/access-om2/issues/129

**Comment by anton-seaice** [comment]
This should be fixed by writing a tool to generate the CICE grid from the MOM-supergrid, which would also give us known provenance of the CICE grid files. (The tool could go in [om3-scripts](https://github.com/COSIMA/om3-scripts)) 

OM3 has some sanity checks (with tolerance set by eps_imesh) but there could be more.

**Comment by anton-seaice** [comment]
@aekiss - I don't have permissions but I think this should go on the CMIP7 project board.

**Comment by anton-seaice** [comment]
The CICE grid.nc file needs these variables defined (per [ice_grid](https://github.com/CICE-Consortium/CICE/blob/64177e3e5658711f5f19030ddfcc6927f1dbda5d/cicecore/cicedyn/infrastructure/ice_grid.F90#L842)) :
```
! (1) ULAT  (radians)    \\
! (2) ULON  (radians)    \\
! (3) HTN   (cm)         \\
! (4) HTE   (cm)         \\
! (7) ANGLE (radians)
```
I think we can derive all these from the mom supergrid.

Angle is defined as 'ANGLE = angle between local x direction and true east' which is consistent with MOM definition except in radians from -pi to pi. Its not clear which corner of the grid cell this is measured, although it is presumably at the U points (for consistency, and because CICE calculates an ANGLET quantity).

**Comment by anton-seaice** [comment]
> Perhaps there was loss of precision due to conversion to single precision and back to double?

Similarly, it could have been a degrees to radian conversion using a constant with insufficient precision

**Comment by aekiss** [comment]
Plots of the grid absolute and relative errors are shown in the first part of https://github.com/aekiss/notebooks-MartinDix/blob/d6d147a/grid025_check.ipynb.

The relative error is very nearly constant (7.746366e-9) for all grids, suggesting a differing value of pi in the deg-rad conversion.

For that matter, it could just be because our numpy check uses a slightly different deg-rad conversion factor from cice.

**Comment by anton-seaice** [comment]
> For that matter, it could just be because our numpy check uses a slightly different deg-rad conversion factor from cice. 

This looks probably ok:

https://github.com/CICE-Consortium/Icepack/blob/f6ff8f7c4d4cb6feabe3651b13204cf43fc948e3/columnphysics/icepack_parameters.F90#L70

`pi     = 3.14159265358979323846_dbl_kind`

```
>>> np.pi
3.141592653589793
```

**Comment by aidanheerdegen** [comment]
So there is a problem, or not?

**Comment by anton-seaice** [comment]
> So there is a problem, or not?

We should fix this, but the discrepancy is <5e-6 so at most the differences in the results would be rounding error ( I think)

We should definitely fix the 0.25 degree grid.

**Comment by anton-seaice** [comment]
I deep dived on tarea too, to see what happens and to try and understand how we should be making the cice c-grid. I calculated it from the mom supergrid like this:

`tarea = mom_ds.area.coarsen(nx=2, ny=2).sum()`

(for sanity, I checked this is the same as `mom_ds.area[::2,::2]+mom_ds.area[1::2,::2]+mom_ds.area[1::2,1::2]+mom_ds.area[::2,1::2]`)

I compared the calculated tarea from the cice file, and the differences south of 65N seem as expected (i.e. in the 14th significant figure).

However, in the tripole region they do become significant (in m^2):
![Screenshot 2024-02-29 at 4 22 47 pm](https://github.com/COSIMA/access-om2/assets/79179784/4d56c7cd-e570-4006-aba0-5f001a7bd24b)

![Screenshot 2024-02-29 at 4 21 10 pm](https://github.com/COSIMA/access-om2/assets/79179784/65e8b151-7541-486f-adf6-aee9bc6af635)

In the cells around the tripole, the difference is >1%. Have I calculated the tarea wrong from the mom grid? Or is the cice file wrong? (The sum of the anomalies is 0)

**Comment by ofa001** [comment]
I have mentioned to @martindix that we did find issues a 1 deg  in strong current in Beaufort Sea was the stongest case, but corner points also were suspect.

**Comment by aekiss** [comment]
FYI the 1deg grid is briefly documented in [Bi and Marsland 2010](https://www.cawcr.gov.au/technical-reports/CTR_027.pdf) and [Bi et al 2013](http://www.bom.gov.au/jshess/docs/2013/bi2_hres.pdf)

**Comment by anton-seaice** [comment]
This issue was fixed here:

https://github.com/COSIMA/esmgrids/commit/1b6837d5fb512adf7f188331fd4dc4b0690bf431

With some testing added too. We should be getting accuracy between mom & cice grids of ~1e-13.

This was added to the OM2 configs here through many PRs linked here: https://github.com/ACCESS-NRI/access-om2-configs/issues/92

**Comment by dougiesquire** [comment]
> I compared the calculated tarea from the cice file, and the differences south of 65N seem as expected (i.e. in the 14th significant figure).

> However, in the tripole region they do become significant (in m^2)

I've been looking into inconsistencies between the ESM1.5 and OM2 grids and I think this difference in areas was actually introduced when the ocean grid was updated to the FMS mosaic format - see [here](https://github.com/ACCESS-NRI/access-esm1.5-configs/issues/23) for details.

**Comment by anton-seaice** [comment]
> I've been looking into inconsistencies between the ESM1.5 and OM2 grids and I think this difference in areas was actually introduced when the ocean grid was updated to the FMS mosaic format - see [here](https://github.com/ACCESS-NRI/access-esm1.5-configs/issues/23) for details.

Is the OM2 ocean supergrid (in the FMS mosaic format?) incorrect then?

**Comment by dougiesquire** [comment]
> Is the OM2 ocean supergrid (in the FMS mosaic format?) incorrect then?

That's what I'm wondering. I need look deeper into how those areas are calculated

**Comment by ofa001** [comment]
Hi @dougiesquire (@anton-seaice), its possible that the precision was increased with access-om2, from the much old access-esm1.5 (originally access-1-0/3 grid) , as you say, as I said when I checked the other day the grid points were nominally the same but the precision as different so yes I can imagine deriver quantities like area of grid boxes will also be impacted.

**Comment by aidanheerdegen** [comment]
> but the precision as different so yes I can imagine deriver quantities like area of grid boxes will also be impacted.

Why only in the tripole? And at the North Pole of a rectilinear grid, so not actually a special location for the tripolar grid?

**Comment by dougiesquire** [comment]
Yeah, I don't think this particular issue is a precision thing. The difference is introduced in the conversion from the old FMS grid format to the mosaic format using FRE-NCtools `transfer_to_mosaic_grid`. It seems that the method that's used to calculate polygon cell areas in FRE-NCtools has an edge-case for polygons (cells) at/near a pole. There's ongoing work in FRE-NCtools to handle this edge case. My best guess at the moment is that our issue is related to that. (See the issue I linked above for more details).

I'll probably just open an issue on FRE-NCtools soon. Hopefully someone there can clarify.

**Comment by aidanheerdegen** [comment]
So that means the old grid has the "more correct" areas?

**Comment by dougiesquire** [comment]
> So that means the old grid has the "more correct" areas?

They're both incorrect, but by slightly different amounts 🙃. If anything, the mosaic format grid (OM2) has the "more correct" areas. The following figure shows the areas along an arc of cells passing through and near the north pole. The bump at the pole shouldn't be there.

<img width="861" alt="Screenshot 2024-06-28 at 1 29 40 PM" src="https://github.com/COSIMA/access-om2/assets/42455466/652609d7-8c42-477f-bf17-ed64ede73da9">

There's an option in the FRE-NCtools polygon area algorithm to calculate the area of polygons near the pole using a copy of the polygon that has been rotated away from the pole. Unfortunately this isn't exposed to `transfer_to_mosaic_grid`, but I added it and that gives the following areas through the pole:

<img width="863" alt="Screenshot 2024-06-28 at 1 33 45 PM" src="https://github.com/COSIMA/access-om2/assets/42455466/54f54e5d-79c1-4f94-acf6-9cfbb3313dcc">

These figures are taken from [this notebook](https://gist.github.com/dougiesquire/23efad4015d97b43e0f319fceb239b2a) which contains more info.

**Comment by ofa001** [comment]
HI @dougiesquire @aidanheerdegen, these grid issues at the pole have been really interesting, one wonders what the implications have been for some of the odd issues we saw at the tripole seem in some of our runs which we never really.
liked. @aekiss looked at them a lot at 0.1 degree but we had them all the way back at 1 deg in ACCESS1-0/3 as well.

**Comment by aekiss** [comment]
There's this probably-unrelated issue which we've never figured out: https://github.com/COSIMA/access-om2/issues/86

**Comment by ofa001** [comment]
Hi @aekiss that's was the link I want you to add in, its probably unrelated but its good to line them up at some point.
